[FLINK-3727] Embedded streaming SQL projection, filtering, union

vasia opened the pull request:
This PR adds support for embedded streaming SQL (projection, filtering, union):
- methods to register DataStreams
- sql translation method in StreamTableEnvironment
- a custom rule to convert to streamable table
- docs for streaming table and sql
- java SQL tests

A streaming SQL query can be executed on a streaming Table by simply adding the `STREAM` keyword in front of the table name. Registering DataStream tables and conversions work in a similar way to that of DataStream tables.
Here's a filtering example:
```
val env = StreamExecutionEnvironment.getExecutionEnvironment
val tEnv = TableEnvironment.getTableEnvironment(env)

val dataStream = env.addSource(...)
val t = dataStream.toTable(tEnv).as('a, 'b, 'c)
tEnv.registerTable("MyTable", t)

val sqlQuery = "SELECT STREAM * FROM MyTable WHERE a = 3"
val result = tEnv.sql(sqlQuery).toDataStream[Row]
```

fhueske commented on docs/apis/batch/libs/table.md:
```diff
@@ -419,8 +438,8 @@ Only the types `LONG` and `STRING` can be casted to `DATE` and vice versa. A `LO
-In order to use a `Table` or `DataSet` in a SQL query, it has to be registered in the `TableEnvironment`, using a unique name.
-A registered `Table` can be retrieved back from the `TableEnvironment` using the `scan()` method:
+In order to use a `Table`, `DataSet`, or `DataStream` in a SQL query, it has to be registered in the `TableEnvironment`, using a unique name.
+A registered Dataset `Table` can be retrieved back from the `TableEnvironment` using the `scan()` method:
```
Add a sentence about the `ingest()` method.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/java/table/StreamTableEnvironment.scala:
```diff
@@ -83,11 +83,54 @@ class StreamTableEnvironment(
   }
 
   /**
+   * Registers the given [[DataStream]] as table in the
```
nitpicking, ScalaDoc indents the `*` under the second `*` of the first line, i.e.,
```
/**
  *
  */
```

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/StreamTableEnvironment.scala:
```diff
@@ -86,6 +86,8 @@ abstract class StreamTableEnvironment(config: TableConfig) extends TableEnvironm
 
     if (isRegistered(tableName)) {
       relBuilder.scan(tableName)
+      //val delta: LogicalDelta = LogicalDelta.create(relBuilder.build())
```
can be removed?

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/StreamTableEnvironment.scala:
```diff
@@ -112,19 +122,31 @@ abstract class StreamTableEnvironment(config: TableConfig) extends TableEnvironm
     * @param name The name under which the table is registered in the catalog.
     * @param dataStream The [[DataStream]] to register as table in the catalog.
+    * @param wrapper True if the registration has to wrap the datastreamTable
+   *                into a [[org.apache.calcite.schema.StreamableTable]]
```
comment indention

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/TableEnvironment.scala:
```diff
@@ -98,8 +98,42 @@ abstract class TableEnvironment(val config: TableConfig) {
+
+  }
+
+  protected def registerStreamTableInternal(name: String, table: AbstractTable): Unit = {
```
same as `registerTableInternal()`, no? So can be removed, IMO.

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/TableEnvironment.scala:
```diff
@@ -98,8 +98,42 @@ abstract class TableEnvironment(val config: TableConfig) {
+   * @param name
+   * @param table
+   */
+  protected def replaceRegisteredStreamTable(name: String, table: AbstractTable): Unit = {
```
rename to `replaceRegisteredTable()` since it is not specific for stream tables?

fhueske commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/sql/streaming/test/FilterITCase.scala:
```diff
@@ -0,0 +1,139 @@
+  }
+
+  @Test
+  def testAllRejectingFilter(): Unit = {
```
This test might be removed.

fhueske commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/sql/streaming/test/FilterITCase.scala:
```diff
@@ -0,0 +1,139 @@
+  }
+
+  @Test
+  def testAllPassingFilter(): Unit = {
```
This test might be removed.

fhueske commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/sql/streaming/test/SelectITCase.scala:
```diff
@@ -0,0 +1,117 @@
+class SelectITCase extends StreamingMultipleProgramsTestBase {
+
+  @Test
+  def testSelectStarFromTable(): Unit = {
```
This test might be removed.

fhueske commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/sql/streaming/test/SelectITCase.scala:
```diff
@@ -0,0 +1,117 @@
+  }
+
+  @Test
+  def testSelectFirstFromTable(): Unit = {
```
This test might be removed.

fhueske commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/sql/streaming/test/SelectITCase.scala:
```diff
@@ -0,0 +1,117 @@
+
+
+  @Test
+  def testSelectSecondFromDataStream(): Unit = {
```
This test might be removed.

fhueske commented:
Thanks for the PR. Looks good. I had a few minor comments. 
I also marked a few tests that could be removed because they do not increase test coverage, IMO. In general, we should be more careful when adding tests. Builds start to fail on Travis because the 2h threshold is exceeded.

vasia commented:
Thank you for the review @fhueske! I've addressed your comments. We can definitely try to figure out a better way to handle the sql tests, as there is a lot of overlap with the table api tests.

fhueske commented:
Thanks for the update @vasia! +1 to merge
